src/modules/article/article.service.ts

```diff
@@ -0,0 +1,253 @@
+        title,
+        description,
+        body,
+        authorId: userId,
+        tagList: {
+          connectOrCreate: tagList.map((tag: string) => ({
+            where: { name: tag },
+            create: { name: tag },
+          })),
+        },
+      },
+      include: {
+        author: {
+          select: {
+            username: true,
+            bio: true,
+            image: true,
+            followers: { select: { id: true } },
+          },
+        },
+        tagList: true,
+        favoritedBy: { select: { id: true } },
+      },
+    });
+
+    return { article: this.buildArticleResponse(article) };
+  }
+
+  async deleteArticle(userId: number, slug: string): Promise<void> {
+    const article = await this.prismaService.article.findUnique({
+      where: { slug },
+      select: {
+        id: true,
+        authorId: true,
+      },
+    });
+
+    if (!article) {
+      throw new NotFoundException(ERROR_ARTICLE_NOT_FOUND);
+    }
+
+    if (article.authorId !== userId) {
+      throw new ForbiddenException(ERROR_FORBIDDEN_DELETE_ARTICLE);
+    }
+
+    await this.prismaService.article.delete({
+      where: { slug },
+    });
+  }
+
+  async updateArticle(
+    userId: number,
+    slug: string,
+    updateArticleDto: UpdateArticleBodyDto,
+  ): Promise<SingleArticleResponse> {
+    const article = await this.prismaService.article.findUnique({
+      where: { slug },
+      select: { id: true, authorId: true },
+    });
+
+    if (!article) {
+      throw new NotFoundException(ERROR_ARTICLE_NOT_FOUND);
+    }
+
+    if (article.authorId !== userId) {
+      throw new ForbiddenException(ERROR_FORBIDDEN_UPDATE_ARTICLE);
+    }
+
+    const data: Partial<UpdateArticleBodyDto & { slug: string }> = {
+      ...updateArticleDto,
+    };
+    if (updateArticleDto.title) {
+      const newSlug = slugify(updateArticleDto.title, { lower: true });
+      if (newSlug !== slug) {
+        const conflictCount = await this.prismaService.article.count({
+          where: { slug: newSlug },
+        });
+
+        if (conflictCount > 0) {
+          throw new ConflictException(ERROR_ARTICLE_CONFLICT);
+        }
+      }
+      data.slug = newSlug;
+    }
+
+    const updatedArticle = await this.prismaService.article.update({
+      where: { slug },
+      data: data,
+      include: {
+        tagList: true,
+        favoritedBy: { select: { id: true } },
+        author: {
+          select: {
+            username: true,
+            bio: true,
+            image: true,
+            followers: { select: { id: true } },
+          },
+        },
+      },
+    });
+
+    return { article: this.buildArticleResponse(updatedArticle, userId) };
+  }
+
+  async listArticles(
+    userId: number | undefined,
+    query: ListArticlesQueryDto,
+  ): Promise<MutlipleArticleResponse> {
+    const { tag, author, favorited } = query;
+    const LIMIT = query.limit ?? DEFAULT_LIMIT;
+    const OFFSET = query.offset ?? DEFAULT_OFFSET;
```

Comment:
chỗ này là khai báo biến rồi, nên đặt tên theo quy ước dành cho biến
```suggestion
    const limit = query.limit ?? DEFAULT_LIMIT;
    const offset = query.offset ?? DEFAULT_OFFSET;
```